packages/eslint-plugin/tests/schema-snapshots/semi.shot

```diff
@@ -0,0 +1,75 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Rule schemas should be convertible to TS types for documentation purposes semi 1`] = `
+"
+# SCHEMA:
+
+{
+  "anyOf": [
+    {
+      "items": [
+        {
+          "enum": ["never"]
+        },
+        {
+          "additionalProperties": false,
+          "properties": {
+            "beforeStatementContinuationChars": {
+              "enum": ["always", "any", "never"]
+            }
+          },
+          "type": "object"
+        }
+      ],
+      "maxItems": 2,
+      "minItems": 0,
+      "type": "array"
+    },
+    {
+      "items": [
+        {
+          "enum": ["always"]
+        },
+        {
+          "additionalProperties": false,
+          "properties": {
+            "omitLastInOneLineBlock": {
+              "type": "boolean"
+            }
+          },
+          "type": "object"
+        }
+      ],
+      "maxItems": 2,
+      "minItems": 0,
+      "type": "array"
+    }
+  ]
+}
+
+
+# TYPES:
+
+type Options =
+  | (
+      | []
+      | ['always']
+      | [
+          'always',
+          {
+            omitLastInOneLineBlock?: boolean;
+          },
+        ]
+    )
+  | (
+      | []
+      | ['never']
+      | [
+          'never',
+          {
+            beforeStatementContinuationChars?: 'always' | 'any' | 'never';
+          },
+        ]
+    );
```

Comment:
do you think its viable to generate array's with `minItems` / `maxItems` to use optional tuple elements?

in some other rules this is getting overly complicated because of that

```ts
type Options =
  | [
      "always"?,
      {
        omitLastInOneLineBlock?: boolean;
      }?
    ]
  | [
      "never"?,
      {
        beforeStatementContinuationChars?: "always" | "any" | "never";
      }?
    ];
```

Reply:
Optional elements are bad because it allows assigning `undefined` - it's less safe.

https://www.typescriptlang.org/play?#code/C4TwDgpgBAKlC8UDaBnYAnAlgOwOYH4AaKbAVwFsAjCdfAXQG4AoJgCiTqhQENhMUAZpggpYASmbsARFM48+g4aJgS2SGcQCMc3vyEjxkpKWwATCEOwRTOhfuWr2J85evFnFnNdt6lhlqCQUACqCMicAD7IaFh4kdEYOLjEZFQ0jCzsPooGwY7qsly6OaJ5RhpQ2kV2fmVqHq421b65+Q1epu5mnlZN8i2lEkA

Reply:
The better solution here would be to improve the algorithm to collapse the union.
EG instead of `([] | [ ..1.. ]) | ([] | [ ..2.. ])`
it should be `[] | [ ..1.. ] | [ ..2.. ]`

For this example it could be:
```ts
type Options =
  | []
  | ['always']
  | [
      'always',
      {
        omitLastInOneLineBlock?: boolean,
      }
    ]
  | ['never']
  | [
      'never',
      {
        beforeStatementContinuationChars?: 'always' | 'any' | 'never',
      }
    ];
```